test/fakeintake/server/server.go

```diff
@@ -41,26 +48,78 @@ func NewServer(port int) *Server {
 	mux := http.NewServeMux()
 	mux.HandleFunc("/", fi.handleDatadogRequest)
 	mux.HandleFunc("/fakeintake/payloads/", fi.getPayloads)
+	mux.HandleFunc("/fakeintake/health/", fi.getFakeHealth)
 
 	fi.server = http.Server{
-		Addr:    fmt.Sprintf(":%d", port),
 		Handler: mux,
+		Addr:    "127.0.0.1:0",
 	}
 
-	fi.start()
+	for _, opt := range options {
+		opt(fi)
+	}
 
 	return fi
 }
 
-func (fi *Server) start() {
+// WithPort changes the server port.
+// If the port is 0, a port number is automatically chosen
+func WithPort(port int) func(*Server) {
+	return func(fi *Server) {
+		if fi.URL() != "" {
+			fmt.Printf("Fake intake is already running. Stop it and try again to change the port.")
+			return
+		}
+		fi.server.Addr = fmt.Sprintf("127.0.0.1:%d", port)
+	}
+}
+
+// WithReadyChannel assign a boolean channel to get notified when the server is ready.
+func WithReadyChannel(ready chan bool) func(*Server) {
+	return func(fi *Server) {
+		if fi.URL() != "" {
+			fmt.Printf("Fake intake is already running. Stop it and try again to change the ready channel.")
+			return
+		}
+		fi.ready = ready
+	}
+}
+
+// Start Starts a fake intake server in a separate go-routine
+// Notifies when ready to the ready channel
+func (fi *Server) Start() {
+	if fi.URL() != "" {
+		fmt.Printf("Fake intake alredy running at %s", fi.URL())
+		return
+	}
 	go func() {
-		err := fi.server.ListenAndServe()
-		if err != nil && err != http.ErrServerClosed {
+		// explicitly creating a listener to get the actual port
+		// as http.Server.ListenAndServe hides this information
+		// https://github.com/golang/go/blob/go1.19.6/src/net/http/server.go#L2987-L3000
+		listener, err := net.Listen("tcp", fi.server.Addr)
+		if err != nil {
 			fmt.Printf("Error creating fake intake server at %s: %v", fi.server.Addr, err)
+
+			fi.ready <- false
+
+			return
+		}
+		fi.url = "http://" + listener.Addr().String()
+		// notify server is ready
+		fi.ready <- true
```

Comment:
What happens if `fi.ready` is `nil`, (e.g. `WithReadyChannel` not called)

Reply:
I thought I was testing this in `server_test`, but with the new public `Start` call I was not going through this code. Sending to a `nil` channel  blocks forever, which is not what we want. I remember trying to check and send only on non `nil` channel, but I was doing something wrong I cannot recall.

This said, changing this to only send on `ready` channel defined, and adding tests to ensure it works as expected

https://go.dev/ref/spec#Send_statements